File: src/cleansia_android/customer-app/src/main/java/cz/cleansia/customer/features/rewards/RewardsActivityViewModel.kt
package cz.cleansia.customer.features.rewards

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import cz.cleansia.core.network.ApiError
import cz.cleansia.core.snackbar.SnackbarController
import cz.cleansia.customer.core.loyalty.LoyaltyActivityItemDto
import cz.cleansia.customer.core.loyalty.LoyaltyRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import javax.inject.Inject
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/**
 * ViewModel backing [RewardsActivityScreen]. Activity history is *not* cached
 * by [LoyaltyRepository] (it's an infrequent drilldown) — this VM owns the
 * paged list locally for the lifetime of the screen.
 *
 * Pagination follows the same shape as [cz.cleansia.customer.core.disputes.DisputeRepository]
 * and OrderRepository: page-1 fetched in `init`, infinite scroll pulls
 * additional pages via [loadNextPage] when the LazyColumn nears its end.
 */
@HiltViewModel
class RewardsActivityViewModel @Inject constructor(
    private val loyaltyRepository: LoyaltyRepository,
    private val snackbar: SnackbarController,
) : ViewModel() {

    private val _items = MutableStateFlow<List<LoyaltyActivityItemDto>>(emptyList())
    val items: StateFlow<List<LoyaltyActivityItemDto>> = _items.asStateFlow()

    private val _loading = MutableStateFlow(false)
    val loading: StateFlow<Boolean> = _loading.asStateFlow()

    private val _loadingMore = MutableStateFlow(false)
    val loadingMore: StateFlow<Boolean> = _loadingMore.asStateFlow()

    private val _loaded = MutableStateFlow(false)
    val loaded: StateFlow<Boolean> = _loaded.asStateFlow()

    private val _total = MutableStateFlow(0)
    val total: StateFlow<Int> = _total.asStateFlow()

    private val pageSize = 20

    init {
        refresh()
    }

    /** Pull-to-refresh — always replaces the list with a fresh page-0 fetch. */
    fun refresh() {
        if (_loading.value) return
        viewModelScope.launch {
            _loading.value = true
            try {
                loyaltyRepository.loadActivity(offset = 0, limit = pageSize)
                    .onSuccess { resp ->
                        _items.value = resp.data
                        _total.value = resp.total
                        _loaded.value = true
                    }
                    .onError { error ->
                        if (error !is ApiError.Network) snackbar.showError(error.getUserMessage())
                    }
            } finally {
                _loading.value = false
            }
        }
    }

    /**
     * Fetch the next page if there are more rows to load. The user can scroll
     * up to retry; a connectivity failure stays silent (NetworkErrorInterceptor
     * owns the infra toast), any other error surfaces the same single message.
     */
    fun loadNextPage() {
        if (_loadingMore.value || _loading.value) return
        if (_items.value.size >= _total.value) return
        viewModelScope.launch {
            _loadingMore.value = true
            try {
                loyaltyRepository.loadActivity(
                    offset = _items.value.size,
                    limit = pageSize,
                )
                    .onSuccess { resp ->
                        _items.value = _items.value + resp.data
                        _total.value = resp.total
                    }
                    .onError { error ->
                        if (error !is ApiError.Network) snackbar.showError(error.getUserMessage())
                    }
            } finally {
                _loadingMore.value = false
            }
        }
    }
}

File: src/cleansia_ios/CleansiaCustomer/Sources/Features/Rewards/RewardsActivityViewModel.swift
import CleansiaCore
import Combine
import Foundation

/// Backs `RewardsActivityScreen` — the full points ledger. Activity isn't cached
/// by `LoyaltyRepository` (an infrequent drilldown), so this VM owns the paged
/// list locally for the screen's lifetime. The Slice-A paged-list archetype:
/// sealed `UiState` + `RefreshPhase`, `refresh()` replaces page 0, `loadNextPage()`
/// appends additively (the `RewardsActivityViewModel.kt` parity).
@MainActor
final class RewardsActivityViewModel: ViewModel {
    @Published private(set) var state: UiState<[LoyaltyActivityItem]> = .loading
    @Published private(set) var refreshPhase: RefreshPhase = .idle
    @Published private(set) var loadingMore = false
    @Published private(set) var hasMore = false

    private let loyaltyRepository: LoyaltyRepository
    private let snackbar: SnackbarController
    private let pageSize: Int
    private var total = 0
    private var loading = false

    init(loyaltyRepository: LoyaltyRepository, snackbar: SnackbarController, pageSize: Int = 20) {
        self.loyaltyRepository = loyaltyRepository
        self.snackbar = snackbar
        self.pageSize = pageSize
    }

    func onAppear() async {
        guard state.loadedValue == nil else { return }
        await refresh()
    }

    func pullToRefresh() async {
        await runRefresh(.userRefreshing)
    }

    func retry() async {
        state = .loading
        await runRefresh(.backgroundRefreshing)
    }

    func refresh() async {
        await runRefresh(.backgroundRefreshing)
    }

    func loadNextPage() async {
        let current = state.loadedValue ?? []
        guard hasMore, !loadingMore, !loading else { return }
        loadingMore = true
        defer { loadingMore = false }
        switch await loyaltyRepository.loadActivity(offset: current.count, limit: pageSize) {
        case let .success(page):
            let combined = current + page.items
            total = page.total
            state = .loaded(combined)
            hasMore = combined.count < total
        case let .failure(error):
            snackbar.showApiError(error)
        }
    }

    private func runRefresh(_ phase: RefreshPhase) async {
        if loading { return }
        loading = true
        refreshPhase = phase
        defer {
            loading = false
            refreshPhase = .idle
        }
        switch await loyaltyRepository.loadActivity(offset: 0, limit: pageSize) {
        case let .success(page):
            total = page.total
            state = .loaded(page.items)
            hasMore = page.items.count < total
        case let .failure(error):
            snackbar.showApiError(error)
            if state.loadedValue == nil {
                state = .error(error)
            }
        }
    }
}
